common/core/web/keyboard-processor/src/keyboards/keyboard.ts

```diff
@@ -98,24 +98,27 @@ namespace com.keyman.keyboards {
     }
 
     /**
-     * HTML help text which is a one liner intended for the status bar of the desktop OSK originally.
+     * HTML help text, as specified by either the &kmw_helptext or &kmw_helpfile system stores.
      * 
-     * Reference: https://help.keyman.com/developer/language/reference/kmw_helptext
+     * Reference: https://help.keyman.com/developer/language/reference/kmw_helptext,
+     *            https://help.keyman.com/developer/language/reference/kmw_helpfile
      */
     get helpText(): string {
       return this.scriptObject['KH'];
     }
 
-    get hasHelpHTML(): boolean {
+    get hasScript(): boolean {
       return !!this.scriptObject['KHF'];
     }
 
     /**
-     * Replaces the OSK with custom HTML, which may be interactive (like with sil_euro_latin).
+     * Embeds a custom script for use by the OSK, which may be interactive (like with sil_euro_latin).
+     * Note:  this must be called AFTER any contents of `helpText` have been inserted into the DOM.
+     * (See sil_euro_latin's source -> sil_euro_latin_js.txt)
```

Comment:
This PR gave me a great opportunity to dive into `sil_euro_latin` and see how it _really_ ticks.

The `KHF` function wraps the keyboard's embedded _script_.

https://github.com/keymanapp/keyboards/blob/1c9365083c10acb758289c5ffb140f5fdc3a4526/release/sil/sil_euro_latin/source/sil_euro_latin.kmn#L15
```
store(&KMW_EMBEDJS) 'sil_euro_latin_js.txt'
```

https://github.com/keymanapp/keyboards/blob/1c9365083c10acb758289c5ffb140f5fdc3a4526/release/sil/sil_euro_latin/source/sil_euro_latin_js.txt#L1-L4

```
this.KHF=function(e) { 
  var h = document.getElementsByTagName('head');
  var stext = 
  "#keyboard_europeanlatin_help { text-align: left; padding: 0; background: #f5e3de; font-size: 8pt; font-family: Tahoma; cursor: default; padding: 4px; }"+
```

Sadly, it seems that Github doesn't like to expand cross-repo permalinks.  ☹️

Reply:
Side-note:  that `KHF` definition has some _nasty_ stuff, some of which will make Shadow DOM tricky to perform for such a "help page" approach in the future.

Like, direct insertion of a style sheet to the <head> element, as seen here.

It also blatantly ignores the provided `e` parameter, preferring to directly retrieve the one it wants via `document.getElementById()` on an element defined in its help-text section.

Reply:
How could we chart a way forward to using shadow dom? What sort of changes would we need to make to keyboards that use kmw_embedjs? (sil_euro_latin, Chinese, Japanese, Korean, perhaps others).

Reply:
Chinese and Japanese are deprecated, .js-only keyboards so far as I can tell.  Neither defines a `KHF` function.  `korean_rr` is not deprecated, and while it does use `kmw_embedjs`, it doesn't actually use `KHF` either.  Help text is available, but it doesn't render a "help page" like `sil_eurolatin` does.

`tibetan_ewts_to_unicode` (also in legacy) is the only other keyboard that defines a `KHF`.

My proposition:

We pass `KHF` a `ShadowRoot` (when possible) specific to the keyboard, rather than the OSK's root `_Box`.

- If necessary (for _old_ browsers / pre-Chromium Edge), JS code can detect whether or not it received an actual shadow root:  https://developer.mozilla.org/en-US/docs/Web/API/Node/getRootNode
- From there:
    1. Build `<style>` element
    2. If ShadowDOM available, insert to shadow root
    3. Else, insert to document (also needed for compat with older KMW versions)
    4. Rinse & repeat with other components.

If the keyboard is able to utilize Shadow DOM stuff, great!  If not... I think we kind of have to support the old, non-shadowed functionality for now - both in keyboards and in the engine.  I don't see any reliable way to shoe-horn non-shadow-aware JS into Shadow DOM stuff.

Reply:
> Chinese and Japanese are deprecated

Legacy, not deprecated. Big difference.